lib/reservoir-sampler.js

```diff
@@ -5,56 +5,68 @@ var es = require('event-stream');
 var _defaults = require('lodash.defaults');
 var debug = require('debug')('mongodb-collection-sample:reservoir-sampler');
 
+var RESERVOIR_SAMPLE_LIMIT = 10000;
 
 /**
- * Take an `_id` and emit the source document.
+ * Take `_id` values, buffer them into a batch, run $in query on a batch
+ * and emit the source documents.
  *
  * @param {mongodb.DB} db
  * @param {String} collectionName to source from.
  * @param {Object} opts
  * @option {Object} fields to return for each document [default: `null`].
  * @return {stream.Transform}
+ *
  * @api private
  */
-function _idToDocument(db, collectionName, opts) {
+function batchDocs(db, collectionName, size, opts) {
   opts = _defaults(opts || {}, {
     fields: null,
+    batchSize: Math.sqrt(size),
```

Comment:
@rueckstiess I think this may hurt us more than help.

I did some testing with `.explain("executionStats")`, and the difference on my local machine between 100 and 1000 _id's was only 7ms.
```
// get some _ids
var ids = db.messages.find({},{_id:1}).limit(1000).toArray()
ids = ids.map(function(el){return el._id})
// test the queries
db.messages.explain("executionStats").find({_id: {$in: ids.slice(0,10)}})
db.messages.explain("executionStats").find({_id: {$in: ids.slice(0,100)}})
db.messages.explain("executionStats").find({_id: {$in: ids.slice(0,1000)}})
```
I think the network round trip will be the real limiting factor here, so we should just get everything in a single batch.